nomad/structs/structs.go

```diff
@@ -5350,6 +5350,34 @@ func (t *Task) Validate(ephemeralDisk *EphemeralDisk, jobType string) error {
 		}
 	}
 
+	// Validation for Kind field which is used for Consul Connect integration
+	// TODO better wording for all error messages
+	if strings.HasPrefix(t.Kind, "connect:") {
+		// This task is a Connect proxy so it should not have service stanzas
+		if len(t.Services) > 0 {
+			mErr.Errors = append(mErr.Errors, fmt.Errorf("Connect proxy task should not have service stanza in it"))
+		}
+		if t.Leader {
+			mErr.Errors = append(mErr.Errors, fmt.Errorf("Connect proxy task must not have leader set"))
+		}
+
+		parts := strings.Split(t.Kind, ":")
+		serviceName := strings.Join(parts[1:], "")
+		if len(parts) > 2 {
+			mErr.Errors = append(mErr.Errors, fmt.Errorf("Proxy service name %q should not contain `:`", serviceName))
+		}
+		// TODO Check if service exists at group level
+		found := false
+		for _, svc := range tgServices {
+			if svc.Name == serviceName {
```

Comment:
Just need to make sure the service has a sidecar specified (although this logic might be better as a small helper method on `type Service`?)

```suggestion
			if svc.Name == serviceName && svc.Connect != nil && svc.Connect.SidecarService != nil {
```